File: services/payment-service/src/models/ledger.model.ts
/**
 * Which step of the saga a journal entry belongs to.
 *
 * FUNDING issues money into the system when a wallet is opened. The other
 * three are the legs of a payment, and a payment's history is readable from
 * this column alone: AUTHORISE + SETTLE completed, AUTHORISE + COMPENSATE was
 * refunded, AUTHORISE on its own is still in flight.
 */
export type Leg = 'FUNDING' | 'AUTHORISE' | 'SETTLE' | 'COMPENSATE';

export interface LedgerEntryRow {
  leg: Leg;
  direction: 'DEBIT' | 'CREDIT';
  amount_cents: number;
  account_id: string;
  /** Joined from accounts, so the audit trail reads without a second lookup. */
  name: string;
  created_at: Date;
}

export interface LedgerEntryDto {
  leg: Leg;
  direction: 'DEBIT' | 'CREDIT';
  amountCents: number;
  accountId: string;
  accountName: string;
  createdAt: Date;
}

export const toLedgerEntryDto = (row: LedgerEntryRow): LedgerEntryDto => ({
  leg: row.leg,
  direction: row.direction,
  amountCents: row.amount_cents,
  accountId: row.account_id,
  accountName: row.name,
  createdAt: row.created_at,
});

/** One line of the general journal, still attached to its group. */
export interface JournalLineRow {
  id: string;
  entry_group: string;
  payment_id: string | null;
  leg: Leg;
  direction: 'DEBIT' | 'CREDIT';
  amount_cents: number;
  created_at: Date;
  account_id: string;
  name: string;
}

export interface JournalLineDto {
  accountId: string;
  accountName: string;
  direction: 'DEBIT' | 'CREDIT';
  amountCents: number;
}

/**
 * A journal entry as a unit: the debit, the credit, and what they were for.
 * Reported rather than assumed to be a pair - a group with any other shape is
 * exactly the corruption the trial balance exists to expose, so it has to be
 * displayable.
 */
export interface JournalEntryDto {
  entryGroup: string;
  paymentId: string | null;
  leg: Leg;
  createdAt: Date;
  amountCents: number;
  lines: JournalLineDto[];
  balanced: boolean;
}

/** Groups flat lines back into journal entries, newest entry first. */
export function toJournalEntries(rows: JournalLineRow[]): JournalEntryDto[] {
  const byGroup = new Map<string, JournalEntryDto>();

  for (const row of rows) {
    let entry = byGroup.get(row.entry_group);
    if (!entry) {
      entry = {
        entryGroup: row.entry_group,
        paymentId: row.payment_id,
        leg: row.leg,
        createdAt: row.created_at,
        amountCents: 0,
        lines: [],
        balanced: false,
      };
      byGroup.set(row.entry_group, entry);
    }
    entry.lines.push({
      accountId: row.account_id,
      accountName: row.name,
      direction: row.direction,
      amountCents: row.amount_cents,
    });
  }

  for (const entry of byGroup.values()) {
    // A debit line first reads the way a journal is written.
    entry.lines.sort((a, b) => (a.direction === 'DEBIT' ? -1 : 1) - (b.direction === 'DEBIT' ? -1 : 1));
    const net = entry.lines.reduce(
      (sum, line) => sum + (line.direction === 'CREDIT' ? line.amountCents : -line.amountCents),
      0,
    );
    entry.balanced = entry.lines.length === 2 && net === 0;
    entry.amountCents = Math.max(...entry.lines.map((line) => line.amountCents));
  }

  return [...byGroup.values()];
}

export interface StatementLineRow {
  entry_group: string;
  payment_id: string | null;
  leg: Leg;
  direction: 'DEBIT' | 'CREDIT';
  amount_cents: number;
  created_at: Date;
  counterparty: string | null;
}

File: services/payment-service/src/main/java/com/ledgerflow/payment/models/LedgerModel.java
package com.ledgerflow.payment.models;

import com.ledgerflow.payment.domain.Direction;
import com.ledgerflow.payment.domain.Leg;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class LedgerModel {

  private LedgerModel() {}

  /** @param accountName joined from accounts, so the audit trail reads without a second lookup. */
  public record LedgerEntryRow(
      Leg leg,
      Direction direction,
      long amountCents,
      String accountId,
      String accountName,
      Instant createdAt) {}

  public record LedgerEntryDto(
      Leg leg,
      Direction direction,
      long amountCents,
      String accountId,
      String accountName,
      Instant createdAt) {}

  public static LedgerEntryDto toLedgerEntryDto(LedgerEntryRow row) {
    return new LedgerEntryDto(
        row.leg(),
        row.direction(),
        row.amountCents(),
        row.accountId(),
        row.accountName(),
        row.createdAt());
  }

  /** One line of the general journal, still attached to its group. */
  public record JournalLineRow(
      long id,
      String entryGroup,
      String paymentId,
      Leg leg,
      Direction direction,
      long amountCents,
      Instant createdAt,
      String accountId,
      String accountName) {}

  public record JournalLineDto(
      String accountId, String accountName, Direction direction, long amountCents) {}

  /**
   * A journal entry as a unit: the debit, the credit, and what they were for.
   * Reported rather than assumed to be a pair - a group with any other shape is
   * exactly the corruption the trial balance exists to expose, so it has to be
   * displayable.
   */
  public record JournalEntryDto(
      String entryGroup,
      String paymentId,
      Leg leg,
      Instant createdAt,
      long amountCents,
      List<JournalLineDto> lines,
      boolean balanced) {}

  /** Groups flat lines back into journal entries, newest entry first. */
  public static List<JournalEntryDto> toJournalEntries(List<JournalLineRow> rows) {
    Map<String, List<JournalLineRow>> byGroup = new LinkedHashMap<>();
    for (JournalLineRow row : rows) {
      byGroup.computeIfAbsent(row.entryGroup(), key -> new ArrayList<>()).add(row);
    }

    List<JournalEntryDto> entries = new ArrayList<>(byGroup.size());
    for (Map.Entry<String, List<JournalLineRow>> group : byGroup.entrySet()) {
      List<JournalLineRow> groupRows = group.getValue();
      JournalLineRow first = groupRows.get(0);

      // A debit line first reads the way a journal is written.
      List<JournalLineRow> ordered = new ArrayList<>(groupRows);
      ordered.sort(Comparator.comparingInt(row -> row.direction() == Direction.DEBIT ? -1 : 1));

      List<JournalLineDto> lines = new ArrayList<>(ordered.size());
      long net = 0;
      long amountCents = 0;
      for (JournalLineRow row : ordered) {
        lines.add(
            new JournalLineDto(
                row.accountId(), row.accountName(), row.direction(), row.amountCents()));
        net += row.direction() == Direction.CREDIT ? row.amountCents() : -row.amountCents();
        amountCents = Math.max(amountCents, row.amountCents());
      }

      entries.add(
          new JournalEntryDto(
              group.getKey(),
              first.paymentId(),
              first.leg(),
              first.createdAt(),
              amountCents,
              List.copyOf(lines),
              lines.size() == 2 && net == 0));
    }

    return List.copyOf(entries);
  }

  public record StatementLineRow(
      String entryGroup,
      String paymentId,
      Leg leg,
      Direction direction,
      long amountCents,
      Instant createdAt,
      String counterparty) {}
}
